File: iosApp/iosApp/Features/NoteDetail/components/comments/CommentItem.swift
//
//  CommentItem.swift
//  iosApp
//
//  Created by velkonost on 09.11.2023.
//  Copyright © 2023 orgName. All rights reserved.
//

import Foundation
import SharedSDK
import SwiftUI

struct CommentItem: View {
    
    let item: Comment
    let onDeleteClick: (Comment) -> Void
    
    init(item: Comment, onDeleteClick: @escaping (Comment) -> Void) {
        self.item = item
        self.onDeleteClick = onDeleteClick
    }
    
    var body: some View {
        VStack(spacing: 0) {
            HStack(spacing: 0) {
                if item.author.avatar != nil {
                    Image(uiImage: UIImage(data: item.author.avatar!.toData()!)!)
                        .resizable()
                        .scaledToFill()
                        .frame(width: 28, height: 28)
                        .clipped()
                        .cornerRadius(8)
                } else {
                    Image(uiImage: SharedR.images().logo.toUIImage()!)
                        .resizable()
                        .frame(width: 28, height: 28)
                        .clipped()
                        .cornerRadius(8)
                }
                
                if item.author.displayName != nil {
                    Text(item.author.displayName!)
                        .style(.bodyLarge)
                        .foregroundColor(.textPrimary)
                        .padding(.leading, 6)
                }
                
                Spacer()
                
                if item.allowEdit {
                    Image(uiImage: SharedR.images().ic_trash.toUIImage()!)
                        .resizable()
                        .renderingMode(.template)
                        .foregroundColor(.iconInactive)
                        .padding(4)
                        .frame(width: 24, height: 24)
                }
            }
            .padding(.top, 6)
            .padding(.horizontal, 4)
            
            HStack {
                Text(item.text)
                    .style(.bodyMedium)
                    .foregroundColor(.textSecondaryTitle)
                    .padding(.top, 12)
                    .padding(.leading, 12)
                Spacer()
            }
            
            HStack {
                Text(item.createdDateStr.localized())
                    .style(.labelSmall)
                    .foregroundColor(.textPrimary)
                    .padding(.top, 6)
                    .padding(.leading, 12)
                Spacer()
            }
        }
//        .padding(.top, 16)
        .padding(.leading, 8)
    
    }
}

File: androidApp/features/notedetail/src/main/kotlin/com/velkonost/getbetter/android/features/notedetail/components/comments/CommentItem.kt
package com.velkonost.getbetter.android.features.notedetail.components.comments

import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.ColorFilter
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import coil.compose.SubcomposeAsyncImage
import coil.request.ImageRequest
import com.velkonost.getbetter.shared.core.model.comments.Comment
import com.velkonost.getbetter.shared.resources.SharedR
import dev.icerock.moko.resources.compose.colorResource
import dev.icerock.moko.resources.compose.painterResource

@Composable
fun CommentItem(
    modifier: Modifier = Modifier,
    item: Comment,
    onDeleteClick: (Comment) -> Unit
) {

    val context = LocalContext.current
    val interactionSource = remember { MutableInteractionSource() }

    Column(
        modifier = modifier
            .fillMaxWidth()
            .padding(top = 16.dp, start = 8.dp)
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            if (item.author.avatar != null) {
                item.author.avatar?.let { avatarBytes ->
                    SubcomposeAsyncImage(
                        modifier = modifier
                            .size(28.dp)
                            .clip(MaterialTheme.shapes.small),
                        model = ImageRequest.Builder(LocalContext.current)
                            .data(
                                BitmapFactory.decodeByteArray(
                                    avatarBytes, 0, avatarBytes.size
                                )
                            )
                            .build(),
                        contentScale = ContentScale.Crop,
                        contentDescription = null,
                    )
                }
            } else {
                Image(
                    modifier = modifier
                        .size(28.dp)
                        .clip(MaterialTheme.shapes.small),
                    painter = painterResource(imageResource = SharedR.images.logo),
                    contentDescription = null
                )
            }

            item.author.displayName?.let {
                Text(
                    modifier = modifier.padding(start = 6.dp),
                    text = it,
                    style = MaterialTheme.typography.bodyLarge,
                    color = colorResource(resource = SharedR.colors.text_primary)
                )
            }

            Spacer(modifier.weight(1f))

            if (item.allowEdit) {
                Image(
                    modifier = modifier
                        .size(24.dp)
                        .padding(4.dp)
                        .clickable(
                            interactionSource = interactionSource,
                            indication = null
                        ) { onDeleteClick.invoke(item) },
                    painter = painterResource(imageResource = SharedR.images.ic_trash),
                    contentDescription = null,
                    colorFilter = ColorFilter.tint(color = colorResource(resource = SharedR.colors.icon_inactive))
                )
            }
        }

        Text(
            modifier = modifier
                .padding(top = 12.dp)
                .padding(start = 12.dp),
            text = item.text,
            style = MaterialTheme.typography.bodyMedium,
            color = colorResource(resource = SharedR.colors.text_secondary)
        )

        Text(
            modifier = modifier
                .padding(top = 6.dp)
                .padding(start = 12.dp),
            text = item.createdDateStr.toString(context),
            style = MaterialTheme.typography.labelSmall,
            color = colorResource(resource = SharedR.colors.text_primary)
        )
    }
}